### PR TITLE
Make remote tagger grpc max message size cofigurable with default of 2MB

### PR DESCRIPTION
**Pull request by adel121**
### What does this PR do?

This PR makes the cluster remote tagger grpc stream max message size configurable with default of 2MB.

### Motivation

Be able to control the max message size using env var. 

Recently, [we did a change](https://github.com/DataDog/datadog-agent/pull/30192) to ensure the message is cut into chunks of 4MB each, however the communication still failed because the size slightly exceeded the 4MB limit due to extra headers added to the message, that's why now we make this parameter configurable with an env var and we use the default value of 2MB to give a sufficient margin for headers.

```
2024-10-22 08:47:22 UTC | CORE | INFO | (comp/core/tagger/taggerimpl/remote/tagger.go:403 in func1) | tagger stream established successfully
2024-10-22 08:47:22 UTC | CORE | WARN | (comp/core/tagger/taggerimpl/remote/tagger.go:307 in run) | error received from remote tagger: rpc error: code = ResourceExhausted desc = trying to send message larger than max (4209972 vs. 4194304)
```

### Describe how to test/QA your changes

No need for QA, E2E tests should be sufficient. (CLC runners should still be able to use remote tagger)

### CODE REVIEW

File: pkg/config/setup/config.go [hunk 1]
```diff
@@ -537,6 +537,7 @@ func InitConfig(config pkgconfigmodel.Setup) {
 	// - nodes
 	config.BindEnvAndSetDefault("cluster_agent.kube_metadata_collection.resources", []string{})
 	config.BindEnvAndSetDefault("cluster_agent.kube_metadata_collection.resource_annotations_exclude", []string{})
+	config.BindEnvAndSetDefault("cluster_agent.cluster_tagger.grpc_max_message_size", 2<<20) // 2 MB
```
Comment (louis-cqrl) [inline pkg/config/setup/config.go:540]: ❓ small question is there a reason going from `4MB` to `2MB` ?

Comment (clamoriniere) [inline pkg/config/setup/config.go:540]: the default value is different from the previous value (2MB vs 4MB) is it expected?

File: cmd/cluster-agent/api/server.go [hunk 2]
```diff
@@ -125,6 +123,7 @@ func StartServer(ctx context.Context, w workloadmeta.Component, taggerComp tagge
 		return struct{}{}, nil
 	})
 
+	maxMessageSize := cfg.GetInt("cluster_agent.cluster_tagger.grpc_max_message_size")
```
Comment (louis-cqrl) [inline cmd/cluster-agent/api/server.go:126]: ❓ don't we want to set a `min` and/or a `max` to the size ?

File: cmd/cluster-agent/api/server.go [hunk 3]
```diff
@@ -134,7 +133,7 @@ func StartServer(ctx context.Context, w workloadmeta.Component, taggerComp tagge
 
 	grpcSrv := grpc.NewServer(opts...)
 	pb.RegisterAgentSecureServer(grpcSrv, &serverSecure{
-		taggerServer: taggerserver.NewServer(taggerComp, maxMessageSize),
+		taggerServer: taggerserver.NewServer(taggerComp, maxMessageSize/2),
```
Comment (clamoriniere) [inline cmd/cluster-agent/api/server.go:136]: this looks like a magic number could you add a comment to explains why you divide by 2 the max message Size here.

File: comp/api/api/apiimpl/server_cmd.go [hunk 4]
```diff
@@ -63,7 +66,7 @@ func (server *apiServer) startCMDServer(
 	pb.RegisterAgentSecureServer(s, &serverSecure{
 		configService:    server.rcService,
 		configServiceMRF: server.rcServiceMRF,
-		taggerServer:     taggerserver.NewServer(server.taggerComp, maxMessageSize),
+		taggerServer:     taggerserver.NewServer(server.taggerComp, maxMessageSize/2),
```
Comment (clamoriniere) [inline comp/api/api/apiimpl/server_cmd.go:69]: same comment about `/2`

**Review by clamoriniere — APPROVED** [review]

File: cmd/cluster-agent/api/server.go [hunk 5]
```diff
@@ -133,8 +132,10 @@ func StartServer(ctx context.Context, w workloadmeta.Component, taggerComp tagge
 	}
 
 	grpcSrv := grpc.NewServer(opts...)
+	// event size should be small enough to fit within the grpc max message size
+	maxEventSize := maxMessageSize / 2
```
Comment (louis-cqrl) [inline cmd/cluster-agent/api/server.go:136]: I think `2` is still a bit arbitrary to me. Isn't a way to compute the size needed dynamically ? @clamoriniere You may have more context than me on this

Comment (clamoriniere) [reply cmd/cluster-agent/api/server.go:136]: We discussed with @adel121, and indeed 2 is a bit arbitrary, however the idea of the message is to container many events, so already 2 Mo is already very large. This setting is to protect the agent against edge cases  but it should never to reach.

@adel121 maybe we could add some observability to monitor the message and even sizes in order to improve later the default max size base on real data.

**Review by louis-cqrl — APPROVED** [review]

**Pull request merged**